pkg/kudoctl/util/repo/package.go

```diff
@@ -0,0 +1,308 @@
+			}
+			paramsStruct := make([]v1alpha1.Parameter, 0)
+			for paramName, param := range params {
+				r := v1alpha1.Parameter{
+					Name:        paramName,
+					Description: param["description"],
+					Default:     param["default"],
+				}
+				paramsStruct = append(paramsStruct, r)
+			}
+			result.Params = paramsStruct
+		default:
+			return fmt.Errorf("unexpected file when reading package from filesystem %s", file.Name())
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
+
+func untarV1Package(r io.Reader) (*v1Package, error) {
+	gzr, err := gzip.NewReader(r)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err := gzr.Close()
+		if err != nil {
+			fmt.Printf("Error when closing gzip reader %s", err)
+		}
+	}()
+
+	tr := tar.NewReader(gzr)
+
+	result := newV1Package()
+	for {
+		header, err := tr.Next()
+
+		switch {
+
+		// if no more files are found return
+		case err == io.EOF:
+			return &result, nil
+
+		// return any other error
+		case err != nil:
+			return nil, err
+
+		// if the header is nil, just skip it (not sure how this happens)
+		case header == nil:
+			continue
+		}
+
+		// check the file type
+		switch header.Typeflag {
+
+		case tar.TypeDir:
+			// we don't need to handle folders, files have folder name in their names and that should be enough
+
+		// if it's a file create it
+		case tar.TypeReg:
+			bytes, err := ioutil.ReadAll(tr)
+			if err != nil {
+				return nil, errors.Wrapf(err, "while reading file from bundle tarball %s", header.Name)
+			}
+
+			switch {
+			case isFrameworkV1File(header.Name):
+				var bf bundle.Framework
+
+				if err = yaml.Unmarshal(bytes, &bf); err != nil {
+					return nil, errors.Wrap(err, "cannot unmarshal framework")
+				}
+				result.Framework = &bf
+			case isTemplateV1File(header.Name):
+				name := strings.TrimPrefix(header.Name, "/templates/")
+				result.Templates[name] = string(bytes)
+			case isParametersV1File(header.Name):
+				var params map[string]map[string]string
+				if err = yaml.Unmarshal(bytes, &params); err != nil {
+					return nil, errors.Wrapf(err, "unmarshalling %s content", header.Name)
+				}
+				paramsStruct := make([]v1alpha1.Parameter, 0)
+				for paramName, param := range params {
+					r := v1alpha1.Parameter{
+						Name:        paramName,
+						Description: param["description"],
+						Default:     param["default"],
+					}
+					paramsStruct = append(paramsStruct, r)
+				}
+				result.Params = paramsStruct
+			default:
+				return nil, fmt.Errorf("unexpected file in the tarball structure %s", header.Name)
+			}
+		}
+	}
+}
+
+func isFrameworkV1File(name string) bool {
+	return strings.HasSuffix(name, frameworkV1FileName)
+}
+
+func isTemplateV1File(name string) bool {
+	matched, _ := regexp.Match(templateV1FileNameRegex, []byte(name))
+	return matched
+}
+
+func isParametersV1File(name string) bool {
+	return strings.HasSuffix(name, paramsV1FileName)
+}
+
+type v1Package struct {
```

Comment:
```suggestion
type V1Package struct {
```

To me its currently unclear what exactly is meant by a V1Package. We are using a lot of terms, `bundle`, `package`, `v1` and `v1alpha` - how do they all tie together?

Is there a difference between a `V1Package` and `V1Bundle` ?

The v1Package contains a Frameworks as a `*bundle.Framework` so whats the difference here?

Reply:
`Is there a difference between a V1Package and V1Bundle ?`

No I am just calling it package, not bundle :) I like that word more and since nobody had any strong opinions there 🤷‍♀